libtcmu.c

```diff
@@ -393,63 +393,186 @@ bool tcmu_dev_oooc_supported(struct tcmu_device* dev)
-	char *reason = NULL;
+	int len, rc;
+	char buf[256], *endbuf;
+	ssize_t ret;
+
+	fd = open(filename, O_RDONLY);
+	if (fd == -1)
+		goto err;
+
+	len = read(fd, buf, sizeof(buf)-1);
+	rc = close(fd);
+	if (len <= 0 || rc == -1)
+		goto err;
+
+	buf[len] = '\0'; /* null-terminate */
+	ret = strtoull(buf, &endbuf, 0);
+	if (buf == endbuf || ret == ULLONG_MAX)
+		goto err;
+
+	return ret;
+
+err:
+	tcmu_warn("cannot read size from %s\n", filename);
+	return -1;
+}
+
+static bool device_parse_cfg(struct tcmu_device *dev,
+			     const char *dev_name, const char *cfgstring)
+{
+	/*
+	 * Extract configuration parameters into dev.
+	 */
 	int len;
+	const char *ptr, *oldptr;
 
-	dev = calloc(1, sizeof(*dev));
-	if (!dev) {
-		tcmu_err("calloc failed in add_device\n");
-		return -ENOMEM;
+	len = snprintf(dev->dev_name, sizeof(dev->dev_name), "%s", dev_name);
+	if (len >= sizeof(dev->dev_name)) {
+		tcmu_err("device name too long for tcmu_device\n");
+		goto err_recompile;
 	}
 
-	snprintf(dev->dev_name, sizeof(dev->dev_name), "%s", dev_name);
-
+	/* Check valid cfgstring */
 	oldptr = cfgstring;
 	ptr = strchr(oldptr, '/');
-	if (!ptr) {
-		tcmu_err("invalid cfgstring\n");
-		goto err_free;
-	}
-
-	if (strncmp(cfgstring, "tcm-user", ptr-oldptr)) {
-		tcmu_err("invalid cfgstring\n");
-		goto err_free;
-	}
+	if (!ptr)
+		goto err_badcfg;
+	if (strncmp(cfgstring, "tcm-user", ptr-oldptr))
+		goto err_badcfg;
 
 	/* Get HBA name */
 	oldptr = ptr+1;
 	ptr = strchr(oldptr, '/');
-	if (!ptr) {
-		tcmu_err("invalid cfgstring\n");
-		goto err_free;
-	}
+	if (!ptr)
+		goto err_badcfg;
 	len = ptr-oldptr;
-	snprintf(dev->tcm_hba_name, sizeof(dev->tcm_hba_name), "user_%.*s", len, oldptr);
+	len = snprintf(dev->tcm_hba_name, sizeof(dev->tcm_hba_name), "user_%.*s", len, oldptr);
+	if (len >= sizeof(dev->tcm_hba_name)) {
+		tcmu_err("hba name too long for tcmu_device\n");
+		goto err_recompile;
+	}
 
 	/* Get device name */
 	oldptr = ptr+1;
 	ptr = strchr(oldptr, '/');
-	if (!ptr) {
-		tcmu_err("invalid cfgstring\n");
-		goto err_free;
-	}
+	if (!ptr)
+		goto err_badcfg;
 	len = ptr-oldptr;
-	snprintf(dev->tcm_dev_name, sizeof(dev->tcm_dev_name), "%.*s", len, oldptr);
+	len = snprintf(dev->tcm_dev_name, sizeof(dev->tcm_dev_name), "%.*s", len, oldptr);
+	if (len >= sizeof(dev->tcm_dev_name)) {
+		tcmu_err("tcm device name too long for tcmu_device\n");
+		goto err_recompile;
+	}
 
 	/* The rest is the handler-specific cfgstring */
 	oldptr = ptr+1;
-	ptr = strchr(oldptr, '/');
-	snprintf(dev->cfgstring, sizeof(dev->cfgstring), "%s", oldptr);
+	len = snprintf(dev->cfgstring, sizeof(dev->cfgstring), "%s", oldptr);
+	if (len >= sizeof(dev->cfgstring)) {
+		tcmu_warn("additional handler cfgstring was truncated\n");
+		/* not a terminal error, but terminate anyway */
+		dev->cfgstring[sizeof(dev->cfgstring)-1] = '\0';
+	}
+	return true;
+
+err_badcfg:
+	tcmu_err("invalid cfgstring: expecting "
+		 "\"tcm-user/<hba_name>/<tcm_device_name>/<handler_name>\"\n");
```

Comment:
I know it goes against the 80 char rule I wrote before, but for strings like this that get printed in log messages you can just keep them as one long string and not break it up:

`tcmu_err("invalid cfgstring: expecting "tcm-user/<hba_name>/<tcm_device_name>/<handler_name>\"\n");`

It seems to make it easier to grep for a log message in the source.